osv2/mget.go

```diff
@@ -0,0 +1,140 @@
+// An error can be returned if
+//
+//   - The request to OpenSearch fails
+//   - The results json cannot be unmarshalled
+func (m *MGetRequest) Do(ctx context.Context, client *opensearch.Client) (*MGetResponse, error) {
+	bodyBytes, jErr := json.Marshal(m)
+	if jErr != nil {
+		return nil, jErr
+	}
+
+	osResp, rErr := opensearchapi.MgetRequest{
+		Index: m.Index,
+		Body:  bytes.NewReader(bodyBytes),
+	}.Do(ctx, client)
+
+	if rErr != nil {
+		return nil, rErr
+	}
+
+	resp := &MGetResponse{
+		StatusCode: osResp.StatusCode,
+		Header:     osResp.Header,
+	}
+
+	var respBuf bytes.Buffer
+	if _, err := respBuf.ReadFrom(osResp.Body); err != nil {
+		return nil, err
+	}
+
+	if err := json.Unmarshal(respBuf.Bytes(), &resp); err != nil {
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+// FromModelMGetRequest creates a new [mgetRequest] from the given [opensearchtools.MGetRequest].
+func FromModelMGetRequest(req *opensearchtools.MGetRequest) *MGetRequest {
+	return &MGetRequest{
+		Index: req.Index,
+		Docs:  req.Docs,
+	}
+}
+
+// MarshalJSON marshals the [MGetRequest] into the proper json expected by OpenSearch 2.
+func (m *MGetRequest) MarshalJSON() ([]byte, error) {
+	docs := make([]any, len(m.Docs))
+	for i, d := range m.Docs {
+		docReq := map[string]any{
+			"_id": d.ID(),
+		}
+
+		if d.Index() != "" {
+			docReq["_index"] = d.Index()
+		}
+
+		docs[i] = docReq
+	}
+
+	source := map[string]any{
+		"docs": docs,
+	}
+
+	return json.Marshal(source)
+}
+
+// MGetResponse is an OpenSearch 2 specific struct corresponding to opensearchapi.Response and [opensearchtools.MGetResponse].
+// It holds a slice of mgetResults.
+type MGetResponse struct {
+	StatusCode int          `json:"-"`
+	Header     http.Header  `json:"-"`
+	Docs       []MGetResult `json:"docs,omitempty"`
+}
+
+// ToModel converts this instance of an [MGetResponse] into an [opensearchtools.MGetResponse].
+func (r *MGetResponse) ToModel() *opensearchtools.MGetResponse {
+	modelDocs := make([]opensearchtools.MGetResult, len(r.Docs))
+	for i, d := range r.Docs {
+		modelDoc := d.ToModel()
+		modelDocs[i] = modelDoc
+	}
+
+	return &opensearchtools.MGetResponse{
+		StatusCode: r.StatusCode,
+		Header:     r.Header,
+		Docs:       modelDocs,
+	}
+}
+
+// mgetResult is the individual result for each requested item.
+type MGetResult struct {
+	Index       string          `json:"_index,omitempty"`
+	ID          string          `json:"_id,omitempty"`
+	Version     int             `json:"_version,omitempty"`
+	SeqNo       int             `json:"_seq_no,omitempty"`
+	PrimaryTerm int             `json:"_primary_term,omitempty"`
+	Found       bool            `json:"found,omitempty"`
+	Source      json.RawMessage `json:"_source,omitempty"`
+	Error       error           `json:"-"`
+}
+
+// ToModel converts this instance of an [MGetResult] into an [opensearchtools.MGetResult].
+func (r *MGetResult) ToModel() opensearchtools.MGetResult {
```

Comment:
There's something that feels lacking to me with just `ToModel`, but I'm not sure if I have an actual answer. Maybe too generic? `ToDomainModel` feels slightly better to me, but maybe domain and model are redundant.

Reply:
Definitely up for a better name, though I agree ToDomainModel doesn't really tell you any more than ToModel.

Reply:
We can punt it until we think of a better name.